### ISSUE TITLE
Removed extension is kept by dev mode after restart

### ISSUE THREAD

**Issue by fedinskiy** [quarkusio/quarkus#34029]
### Describe the bug

I have an application, running in dev-mode. When I remove one of the extensions from the pom, dev-mode restarts, but the extension is still considered installed.

### Expected behavior

Running application should reflect changes in the pom.xml, extension removed from the pom should not be used in the app.

### Actual behavior

Extension is keep until manual restart

### How to Reproduce?

1. ` quarkus create app app --platform-bom=io.quarkus:quarkus-bom:3.1.0.Final -x=quarkus-smallrye-health,quarkus-resteasy-reactive`
2. `cd app`
3. ` quarkus dev -Dquarkus.platform.version=999-SNAPSHOT -Dquarkus.platform.artifact-id=quarkus-bom -Dquarkus.http.port=1101 -Dquarkus.platform.group-id=io.quarkus` 
4. The log contains `Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, smallrye-health, vertx]`, command ` curl localhost:1101/q/health` returns health status.
5. Run `quarkus extension remove quarkus-smallrye-health` from another console(or remove the extension manually from the pom.xml), result: `[SUCCESS] ✅  Extension io.quarkus:quarkus-smallrye-health has been uninstalled`, pom doesn't contain `smallrye-health` anymore
6. Logs of the running app: 
```
[INFO] Changes detected to [/home/fedinskiy/code/quarkus/refresh/app/pom.xml], restarting dev mode
<...>
 Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, smallrye-health, vertx]
```
command ` curl localhost:1101/q/health` returns health status.
7. Stop and start app manually. Log `Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]`. Running curl over the health endpoint returns 404, as it should.

### Output of `uname -a` or `ver`

4.18.0-425.19.2.el8_7.x86_64

### Output of `java -version`

17.0.5, vendor: GraalVM Community

### Quarkus version or git rev

d60dea5d8c3e443e52ce859f2bb279ca96385db5

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)

### Additional information

```
$ quarkus --version
2.14.3.Final # can be reproduced with 3.1.0.Final as well
```

**Comment by rsvoboda** [comment]
io.quarkus:quarkus-bom:3.0.0.Alpha4 @fedinskiy ?

**Comment by fedinskiy** [comment]
@rsvoboda it can be reproduced with 3.1.0 for both as well. Added more details to the description

**Comment by michalvavrik** [comment]
3.0.0.Alpha4 is just random old community version, we can raise it, but it is not related as @fedinskiy said.

**Comment by rsvoboda** [comment]
CC @maxandersen @aloubyansky @ia3andy

**Comment by aloubyansky** [comment]
Worked for me on first attempt
```
[INFO] --- quarkus:999-SNAPSHOT:dev (default-cli) @ code-with-quarkus ---
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:3.3.0:resources (default-resources) @ code-with-quarkus
[INFO] Copying 2 resources
[INFO] Invoking io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code (default) @ code-with-quarkus
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) @ code-with-quarkus
[INFO] Changes detected - recompiling the module! :source
[INFO] Compiling 2 source files with javac [debug release 17] to target/classes
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:3.3.0:testResources (default-testResources) @ code-with-quarkus
[INFO] skip non existing resourceDirectory /home/aloubyansky/playground/code-with-quarkus/src/test/resources
[INFO] Invoking io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code-tests (default) @ code-with-quarkus
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) @ code-with-quarkus
[INFO] Changes detected - recompiling the module! :dependency
[INFO] Compiling 2 source files with javac [debug release 17] to target/test-classes
Listening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2023-06-30 18:32:15,482 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.532s. Listening on: http://localhost:8080

2023-06-30 18:32:15,486 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-06-30 18:32:15,486 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, smallrye-health, vertx]

--
Tests paused
Press [e] to edit command line args (currently ''), [r] to resume testing, [o] Toggle test output, [:] for the terminal, [h] for more options>[INFO] Changes detected to [/home/aloubyansky/playground/code-with-quarkus/pom.xml], restarting dev mode
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:3.3.0:resources (default-resources) @ code-with-quarkus
[INFO] Copying 2 resources
[INFO] Invoking io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code (default) @ code-with-quarkus
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) @ code-with-quarkus
[INFO] Nothing to compile - all classes are up to date
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:3.3.0:testResources (default-testResources) @ code-with-quarkus
[INFO] skip non existing resourceDirectory /home/aloubyansky/playground/code-with-quarkus/src/test/resources
[INFO] Invoking io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code-tests (default) @ code-with-quarkus
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.11.0:testCompile (default-testCompile) @ code-with-quarkus
[INFO] Changes detected - recompiling the module! :dependency
[INFO] Compiling 2 source files with javac [debug release 17] to target/test-classes

Listening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2023-06-30 18:32:47,032 INFO  [io.quarkus] (Quarkus Main Thread) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.434s. Listening on: http://localhost:8080

2023-06-30 18:32:47,035 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-06-30 18:32:47,036 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, resteasy-reactive, smallrye-context-propagation, vertx]
```

**Comment by aloubyansky** [comment]
Can't seem to reproduce it so far

**Comment by fedinskiy** [comment]
@aloubyansky is it possible, that you use very old 999 build?
I was able to reproduce the error on the fresh machine using `quarkus dev -Dquarkus.platform.version=3.2.0.CR1 -Dquarkus.platform.artifact-id=quarkus-bom -Dquarkus.http.port=1101 -Dquarkus.platform.group-id=io.quarkus` for step 3.
If you still have problems I could try to recheck the latest snapshot on Monday.

```
Apache Maven 3.8.8 (4c87b05d9aedce574290d1acc98575ed5eb6cd39)
Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: 22.3.1.r17-mandrel
OS name: "linux", version: "5.14.21-150500.53-default", arch: "amd64", family: "unix"

```

**Comment by aloubyansky** [comment]
I can reproduce it with 3.2.0.CR1 but not 3.2.0.Final. I used the main branch before. Looks like one of my recent PRs has fixed it already.

**Comment by aloubyansky** [comment]
It was probably https://github.com/quarkusio/quarkus/pull/34286

**Comment by fedinskiy** [comment]
I confirm, that this is fixed in (at least) 3aa2fbbcb60b40c850d7baccbe7cdc295a9b5b85. Thank you, @aloubyansky !
